src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +23,108 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return BATTERY_AMP_HOUR_CAPACITY-ampHours ;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(newCurrent<0)
+    {
     return false;
+    }
+    else
+    {
+       return true;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return t;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
     Q_UNUSED(batteryData);
     // Update your variables here.
+    initialCurrent = newCurrent;
+    newCurrent = batteryData.current;
+    if(initialCurrent == 0)
+    {
+        aveCurrent = newCurrent;
+    }
+    else
+    {
+        aveCurrent = (initialCurrent+newCurrent)/2;
+    }
+    initialTime = newTime;
+    newTime = batteryData.time;
+
+    dAmpHours = aveCurrent*dTime;
+    //qDebug() << "new time" << newTime;
+    //qDebug() << "initial time" <<initialTime;
+    dTime = initialTime.msecsTo(newTime);
+
+    dTime = dTime*2.77778e-7;
```

Comment:
This number `2.77778e-7` deserves to be a named constant, make the name something that represents where this value came from. A common term for these number that appear in the code are "magic numbers" (http://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad).